File: Pixler.py
import socket


class Pixler:

    def __init__(self, host, port):
        """
        initialize the server connection
        :param host: the IP address of the pixelflut Server
        :param port: the port of the pixelflut server
        """
        self.HOST = host
        self.PORT = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((host, port))

    def send(self, message):
        """
        send the given message to the server.\n \n
        supported commands:
            - `HELP`: Prints a help text with the available commands. \n
            - `PX x y rrggbb`: PX x y rrggbb: Color the pixel (x,y) with the given hexadecimal color rrggbb, e.g. PX 10 10 ff0000 \n
            - `PX x y rrggbbaa`: Color the pixel (x,y) with the given hexadecimal color rrggbb (alpha channel is ignored for now), e.g. PX 10 10 ff0000ff \n
            - `PX x y gg`: Color the pixel (x,y) with the hexadecimal color gggggg. Basically this is the same as the other commands, but is a more efficient way of filling white, black or gray areas, e.g. PX 10 10 00 to paint black \n
            - `PX x y`: Get the color value of the pixel (x,y), e.g. PX 10 10 \n
            - `SIZE`: Get the size of the drawing surface, e.g. SIZE 1920 1080 \n
            - `OFFSET x y`: Apply offset (x,y) to all further pixel draws on this connection. This can e.g. be used to pre-calculate an image/animation and simply use the OFFSET command to move it around the screen without the need to re-calculate it, e.g. OFFSET 100 100 \n

        :param message: the message to send to the sercer e.g. 'PX x y rrggbb'
        :return: nothing
        """
        self.sock.sendto(message.encode(), (self.HOST, self.PORT))

    def send_and_receive(self, message):
        """
        same as the send-methode but returns the result
        :param message: the message to send
        :return: the response from the server
        """
        self.sock.sendall(message.encode())
        response = self.sock.recv(1024).decode()
        # print(response)
        return response

    def get_size(self):
        """
        request the size from the server and return it as x, y tuple
        :return: the canvas size as (x, y)
        """
        size = self.send_and_receive("SIZE").split()
        x = int(size[1])
        y = int(size[2])
        return x, y

    def set_pixel(self, x, y, r, g, b, a=255):
        """
        set one pixel on the canvas
        :param x: the x coordinate on the canvas
        :param y: the y coordinate on the canvas
        :param r: the red-color value
        :param g: the green-color value
        :param b: the blue-color value
        :param a: the alpha value
        :return: nothing
        """
        if a == 255:
            self.send('PX %d %d %02x%02x%02x\n' % (x, y, r, g, b))
        else:
            self.send('PX %d %d %02x%02x%02x%02x\n' % (x, y, r, g, b, a))

    def get_pixel(self, x, y):
        """
        get the color value for the given pixel
        @:param x the x-coordinate
        @:param y the y-coordinate
        @:returns the color value in hex e.g "c837ff"
        """
        pixel = self.send_and_receive("PX %d %d\n" % (x, y)).split()
        color = pixel[3]
        return color

File: java/src/Pixler.java
import java.awt.*;
import java.net.*;
import java.io.*;

public class Pixler {
    private Socket clientSocket;
    private  PrintWriter out;
    private  BufferedReader in;

    private int[] size;

    public Pixler(String host, int port) {
        try {
            clientSocket = new Socket(host, port);
            System.out.println("Socket successfully created and connected to the server.");
            out = new PrintWriter(clientSocket.getOutputStream(), false);
            in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
            size = getSize();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
    }

    /**
     * send one message to the server
     *  supported commands:
     *  <p>
     * - `HELP`: Prints a help text with the available commands. <br>
     * - `PX x y rrggbb`: PX x y rrggbb: Color the pixel (x,y) with the given hexadecimal color rrggbb, e.g. PX 10 10 ff0000 <br>
     * - `PX x y rrggbbaa`: Color the pixel (x,y) with the given hexadecimal color rrggbb (alpha channel is ignored for now), e.g. PX 10 10 ff0000ff <br>
     * - `PX x y gg`: Color the pixel (x,y) with the hexadecimal color gggggg. Basically this is the same as the other commands, but is a more efficient way of filling white, black or gray areas, e.g. PX 10 10 00 to paint black <br>
     * - `PX x y`: Get the color value of the pixel (x,y), e.g. PX 10 10 <br>
     * - `SIZE`: Get the size of the drawing surface, e.g. SIZE 1920 1080 <br>
     * - `OFFSET x y`: Apply offset (x,y) to all further pixel draws on this connection. This can e.g. be used to pre-calculate an image/animation and simply use the OFFSET command to move it around the screen without the need to re-calculate it, e.g. OFFSET 100 100 <br>
     *  </p>
     *  @param message the message to send
     */
    public void sendMessage (String message) {
        out.println(message);
    }

    /**
     * send a message to the server and wait for the response
     * @param message the message to send
     * @return the answer of the server
     */
    public String sendMessageAndReceive (String message) {
        try {
            out.flush();
            out.println(message);
            out.flush();

            return in.readLine();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        return  null;
    }


    /**
     * close socket connection
     */
    public void terminateConnection() {
        try {
            in.close();
            out.close();
            clientSocket.close();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
    }

    /**
     * set one pixel at position x y on the canvas to the color @color
     * @param x the x-coordinate on the canvas
     * @param y the y-coordinate on the canvas
     * @param color the color value for the pixel
     */
    public void setPixel(int x, int y, Color color) {
        // System.out.printf("PX %d %d %02x%02x%02x\\n", x, y, color.getRed(), color.getGreen(), color.getBlue());
        sendMessage(String.format("PX %d %d %02x%02x%02x\\n", x, y, color.getRed(), color.getGreen(), color.getBlue()));
    }

    /**
     * get the color value of the pixel at x, y
     * @param x the x-coordinate on the canvas
     * @param y the y-coordinate on the canvas
     * @return the color of the pixel
     */
    public Color getPixel(int x, int y) {
        if(x < size[0] & y < size[1]) {
            String color = sendMessageAndReceive(String.format("PX %d %d", x, y)).split(" ")[3];
            System.out.println(color);
            int red = Integer.valueOf(color.substring(0, 2), 16);
            int green = Integer.valueOf(color.substring(2, 4), 16);
            int blue = Integer.valueOf(color.substring(4, 6), 16);
            return new Color(red, green, blue);
        } else {
            System.err.printf("Pixel %d %d ist außerhalb des Bildbereichs (%d x %d) \n", x, y, size[0], size[1]);
        }
        return  null;
    }

    /**
     * get the size of the canvas
     * @return array with the size of the canvas [x, y]
     */
    public int[] getSize() {
        int[] result = new int[2];
        String[] size = sendMessageAndReceive("SIZE").split(" ");
        result[0] = Integer.parseInt(size[1]);
        result[1] = Integer.parseInt(size[2]);
        return result;
    }
}
